File: Core/PythiaCore/Sources/PythiaCore/PortableBackup.swift
import Foundation

public struct PythiaPortableSettings: Codable, Equatable, Sendable {
    public var sourceLanguage: String?
    public var targetLanguage: String?
    public var enabledTranslateServices: [String]?
    public var translateServiceOrder: [String]?
    public var openAICompatibleEnabled: Bool?
    public var openAICompatibleName: String?
    public var openAICompatibleBaseUrl: String?
    public var openAICompatibleModel: String?
    public var deepLEnabled: Bool?
    public var deepLBaseUrl: String?
    public var libreTranslateEnabled: Bool?
    public var libreTranslateBaseUrl: String?
    public var saveHistory: Bool?
    public var themeMode: String?

    public init(
        sourceLanguage: String? = nil,
        targetLanguage: String? = nil,
        enabledTranslateServices: [String]? = nil,
        translateServiceOrder: [String]? = nil,
        openAICompatibleEnabled: Bool? = nil,
        openAICompatibleName: String? = nil,
        openAICompatibleBaseUrl: String? = nil,
        openAICompatibleModel: String? = nil,
        deepLEnabled: Bool? = nil,
        deepLBaseUrl: String? = nil,
        libreTranslateEnabled: Bool? = nil,
        libreTranslateBaseUrl: String? = nil,
        saveHistory: Bool? = nil,
        themeMode: String? = nil
    ) {
        self.sourceLanguage = sourceLanguage
        self.targetLanguage = targetLanguage
        self.enabledTranslateServices = enabledTranslateServices
        self.translateServiceOrder = translateServiceOrder
        self.openAICompatibleEnabled = openAICompatibleEnabled
        self.openAICompatibleName = openAICompatibleName
        self.openAICompatibleBaseUrl = openAICompatibleBaseUrl
        self.openAICompatibleModel = openAICompatibleModel
        self.deepLEnabled = deepLEnabled
        self.deepLBaseUrl = deepLBaseUrl
        self.libreTranslateEnabled = libreTranslateEnabled
        self.libreTranslateBaseUrl = libreTranslateBaseUrl
        self.saveHistory = saveHistory
        self.themeMode = themeMode
    }
}

public struct PythiaPortableBackup: Codable, Equatable, Sendable {
    public static let currentSchemaVersion = 1

    public var schemaVersion: Int
    public var product: String
    public var createdAt: Date
    public var sensitiveFieldsOmitted: Bool
    public var settings: PythiaPortableSettings
    public var history: [PythiaHistoryRecord]

    public init(
        schemaVersion: Int = PythiaPortableBackup.currentSchemaVersion,
        product: String = "Pythia",
        createdAt: Date = Date(),
        sensitiveFieldsOmitted: Bool = true,
        settings: PythiaPortableSettings,
        history: [PythiaHistoryRecord]
    ) {
        self.schemaVersion = schemaVersion
        self.product = product
        self.createdAt = createdAt
        self.sensitiveFieldsOmitted = sensitiveFieldsOmitted
        self.settings = settings
        self.history = history
    }
}

public enum PythiaPortableBackupError: LocalizedError, Equatable {
    case unsupportedSchema(Int)
    case foreignProduct(String)
    case unsafePayload

    public var errorDescription: String? {
        switch self {
        case let .unsupportedSchema(version):
            return "不支持 Pythia 备份格式版本 \(version)。"
        case let .foreignProduct(product):
            return "这不是 Pythia 备份文件（product: \(product)）。"
        case .unsafePayload:
            return "备份未声明已排除敏感字段。"
        }
    }
}

public enum PythiaPortableBackupCodec {
    public static func encode(_ backup: PythiaPortableBackup) throws -> Data {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        return try encoder.encode(backup)
    }

    public static func decode(_ data: Data) throws -> PythiaPortableBackup {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        let backup = try decoder.decode(PythiaPortableBackup.self, from: data)
        guard backup.schemaVersion == PythiaPortableBackup.currentSchemaVersion else {
            throw PythiaPortableBackupError.unsupportedSchema(backup.schemaVersion)
        }
        guard backup.product == "Pythia" else {
            throw PythiaPortableBackupError.foreignProduct(backup.product)
        }
        guard backup.sensitiveFieldsOmitted else {
            throw PythiaPortableBackupError.unsafePayload
        }
        return backup
    }
}

File: Windows/Pythia.Windows/lib/core/portable_backup.dart
import 'dart:convert';

import 'history_record.dart';
import 'history_sync.dart';
import 'settings_model.dart';
import 'webdav_sync.dart';

class PortableBackupRestoreResult {
  final PythiaSettings settings;
  final int importedHistoryCount;
  final int mergedHistoryCount;
  final int conflictCount;

  const PortableBackupRestoreResult({
    required this.settings,
    required this.importedHistoryCount,
    required this.mergedHistoryCount,
    required this.conflictCount,
  });
}

class PortableBackupService {
  static const schemaVersion = 1;

  static const _portableSettingKeys = <String>{
    'sourceLanguage',
    'targetLanguage',
    'enabledTranslateServices',
    'translateServiceOrder',
    'openAICompatibleEnabled',
    'openAICompatibleName',
    'openAICompatibleBaseUrl',
    'openAICompatibleModel',
    'deepLEnabled',
    'deepLBaseUrl',
    'libreTranslateEnabled',
    'libreTranslateBaseUrl',
    'saveHistory',
    'themeMode',
  };

  final HistoryRepository historyRepository;

  const PortableBackupService(this.historyRepository);

  Future<String> create(PythiaSettings settings) async {
    final allSettings = settings.toJson();
    final portableSettings = <String, Object?>{
      for (final key in _portableSettingKeys) key: allSettings[key],
    };
    final history = await historyRepository.readAllForSync();
    return const JsonEncoder.withIndent('  ').convert({
      'schemaVersion': schemaVersion,
      'product': 'Pythia',
      'createdAt': DateTime.now().toUtc().toIso8601String(),
      'sensitiveFieldsOmitted': true,
      'settings': portableSettings,
      'history': history.map((record) => record.toJson()).toList(),
    });
  }

  Future<PortableBackupRestoreResult> restore(
    String encoded,
    PythiaSettings currentSettings,
  ) async {
    final Object? decoded;
    try {
      decoded = jsonDecode(encoded);
    } catch (error) {
      throw FormatException('备份不是有效的 JSON：$error');
    }
    if (decoded is! Map<String, Object?>) {
      throw const FormatException('备份根节点必须是对象。');
    }
    if (decoded['product'] != 'Pythia') {
      throw const FormatException('这不是 Pythia 备份文件。');
    }
    if (decoded['schemaVersion'] != schemaVersion) {
      throw const FormatException('不支持该备份格式版本。');
    }
    final rawSettings = decoded['settings'];
    final rawHistory = decoded['history'];
    if (rawSettings is! Map<String, Object?> || rawHistory is! List<Object?>) {
      throw const FormatException('备份缺少有效的设置或历史记录。');
    }

    final portableSettings = <String, Object?>{
      for (final entry in rawSettings.entries)
        if (_portableSettingKeys.contains(entry.key)) entry.key: entry.value,
    };
    final importedServiceIds =
        (portableSettings['enabledTranslateServices'] as List<Object?>?)
                ?.whereType<String>()
                .toSet() ??
            const <String>{};
    if (portableSettings.containsKey('enabledTranslateServices')) {
      portableSettings['googleEnabled'] =
          importedServiceIds.contains(PythiaSettings.googleServiceId);
      portableSettings['baiduEnabled'] =
          importedServiceIds.contains(PythiaSettings.baiduServiceId);
      portableSettings['youdaoEnabled'] =
          importedServiceIds.contains(PythiaSettings.youdaoServiceId);
    }
    final decodedSettings = PythiaSettings.fromJson({
      ...currentSettings.toJson(),
      ...portableSettings,
    }).normalized();
    final availableServices = <String>{
      'local',
      if (decodedSettings.googleEnabled) PythiaSettings.googleServiceId,
      if (decodedSettings.baiduEnabled) PythiaSettings.baiduServiceId,
      if (decodedSettings.youdaoEnabled) PythiaSettings.youdaoServiceId,
      if (decodedSettings.openAICompatibleEnabled)
        PythiaSettings.openAICompatibleServiceId,
      if (decodedSettings.deepLEnabled) PythiaSettings.deepLServiceId,
      if (decodedSettings.libreTranslateEnabled)
        PythiaSettings.libreTranslateServiceId,
    };
    final enabledServices = decodedSettings.enabledTranslateServices
        .where(availableServices.contains)
        .toList();
    if (enabledServices.isEmpty) enabledServices.add('local');
    final serviceOrder = <String>[
      for (final id in decodedSettings.translateServiceOrder)
        if (enabledServices.contains(id)) id,
      for (final id in enabledServices)
        if (!decodedSettings.translateServiceOrder.contains(id)) id,
    ];
    final restoredSettings = decodedSettings.copyWith(
      enabledTranslateServices: enabledServices,
      translateServiceOrder: serviceOrder,
    );
    final imported = <PythiaHistoryRecord>[];
    try {
      for (final item in rawHistory) {
        if (item is! Map<String, Object?>) {
          throw const FormatException('历史记录条目格式错误。');
        }
        imported.add(PythiaHistoryRecord.fromJson(item));
      }
    } catch (error) {
      if (error is FormatException) rethrow;
      throw FormatException('历史记录解析失败：$error');
    }

    final local = await historyRepository.readAllForSync();
    final merged = PythiaHistoryMerger.merge(local: local, remote: imported);
    final pending = [
      for (final record in merged.merged)
        if (record.syncStatus == PythiaSyncStatus.conflict)
          record
        else
          record.copyWith(
            syncStatus: record.isDeleted
                ? PythiaSyncStatus.pendingDelete
                : PythiaSyncStatus.pendingUpload,
          ),
    ];

    await historyRepository.backupBeforeSync();
    await historyRepository.replaceAllFromSync(pending);
    return PortableBackupRestoreResult(
      settings: restoredSettings,
      importedHistoryCount: imported.length,
      mergedHistoryCount: pending.length,
      conflictCount: merged.conflicts.length,
    );
  }
}
